File: dance-cognition/connectivity/event_packing.py
import json
from event import Event
import numpy

class EventPacker:
    @classmethod
    def pack(cls, event):
        return json.dumps(cls._serialize(event))

    @classmethod
    def unpack(cls, string):
        return json.loads(string, object_hook=cls._restore)

    @classmethod
    def _serialize(cls, obj):
        if obj is None or isinstance(obj, (bool, int, long, float, basestring)):
            return obj
        if isinstance(obj, list):
            return [cls._serialize(val) for val in obj]
        if isinstance(obj, dict):
            return {"py/dict": dict(
                    (k, cls._serialize(v))
                    for k, v in obj.iteritems()
                    )}
        if isinstance(obj, numpy.ndarray):
            return {
                "py/numpy.ndarray": {
                    "values": obj.tolist(),
                    "dtype":  str(obj.dtype)}}
        if isinstance(obj, Event):
            return {
                "py/Event": {
                    "type": obj.type,
                    "content": cls._serialize(obj.content),
                    "source": obj.source}}
        raise TypeError("Type %s not obj-serializable" % type(obj))

    @classmethod
    def _restore(cls, obj):
        if not isinstance(obj, dict):
            return obj

        try:
            return dict(obj["py/dict"])
        except KeyError:
            pass

        try:
            data = obj["py/numpy.ndarray"]
            return numpy.array(data["values"], dtype=data["dtype"])
        except KeyError:
            pass

        try:
            data = obj["py/Event"]
            event = Event(data["type"], cls._restore(data["content"]))
            event.source = data["source"]
            return event
        except KeyError:
            pass

        return obj

File: dance-cognition/experiments/dimensionality_reduction/html5/event_packing.js
function packEvent(event) {
    return JSON.stringify(serialize(event));
}

function serialize(event) {
    if(event.content instanceof PyDict)
	content = {"py/dict": event.content.dict};
    else
	content = event.content;

    return {"py/Event": {
	"type": event.type,
	"content": content,
	"source": event.source
    }};
}

function unpackEvent(string) {
    return unserialize(JSON.parse(string));
}

function unserialize(obj) {
    var eventObj = obj["py/Event"];
    if(eventObj) {
	var type = eventObj.type;
    	if(type == "PARAMETER") {
    	    var content = eventObj.content["py/dict"];
	    return new Event(type, content);
    	}
    }
}
